openstack_auth/utils.py

```diff
@@ -242,10 +265,17 @@ def get_websso_url(request, auth_url, websso_auth):
                '/protocols/%s/websso?origin=%s' %
                (auth_url, idp_id, protocol_id, origin))
     else:
-        # If no IDP mapping found for the identifier,
-        # perform WebSSO by protocol.
-        url = ('%s/auth/OS-FEDERATION/websso/%s?origin=%s' %
-               (auth_url, protocol_id, origin))
+        if (is_websso_default_redirect() and get_websso_default_redirect_url()):
+            # If the IdP is not found (or was explicitly undefined), and
+            # there is a default URL set, prefer it over WebSSO by protocol.
+            host = request.get_host()
+            url = ('%s?origin=%s&host=%s' %
```

Comment:
Note: the `host` part is ONLY so Portal doesn't have to be updated. If we update Portal to accept either `host` or `origin` (preferring origin, which is more precise), then we can drop this.

Reply:
The host param is a callback of sorts, used to verify the host is trusted and in conjunction with `next` lets us redirect users to a specific page in horizon. It looks like this functionality is lost, at least with portal sso and these new changes, can we still deep-link to a specific page like `/project/images`?

Reply:
Good call, I knew there was built-in `next` support in openstack_auth but I wasn't sure how it worked and didn't test it, thanks for the reminder. It looks like it's indeed not working, I'll have to see how to support it.

Reply:
Fixed in 0327103